pinot-core/src/main/java/org/apache/pinot/core/query/optimizer/filter/NumericalFilterOptimizer.java

```diff
@@ -0,0 +1,273 @@
+        // Rewrite the expression.
+        return rewrite(expression, lhs, rhs, schema);
+      }
+    }
+
+    return expression;
+  }
+
+  /**
+   * If any of the operands of AND function is "false", then the AND function itself is false and can be replaced with
+   * "false" literal. Otherwise, remove all the "true" operands of the AND function. Similarly, if any of the operands
+   * of OR function is "true", then the OR function itself is true and can be replaced with "true" literal. Otherwise,
+   * remove all the "false" operands of the OR function.
+   */
+  private static Expression optimizeCurrent(Expression expression) {
+    Function function = expression.getFunctionCall();
+    List<Expression> operands = function.getOperands();
+    if (function.getOperator().equals(FilterKind.AND.name())) {
+      // If any of the literal operands are FALSE, then replace AND function with FALSE.
+      if (operands.stream().anyMatch(operand -> operand.equals(FALSE))) {
+        return setExpressionToBoolean(expression, false);
+      }
+
+      // Remove all Literal operands that are TRUE.
+      operands.removeIf(x -> x.equals(TRUE));
+      if (operands.size() == 0) {
+        return setExpressionToBoolean(expression, true);
+      }
+    } else if (function.getOperator().equals(FilterKind.OR.name())) {
+      // If any of the literal operands are TRUE, then replace OR function with TRUE
+      if (operands.stream().anyMatch(operand -> operand.equals(TRUE))) {
+        return setExpressionToBoolean(expression, true);
+      }
+
+      // Remove all Literal operands that are FALSE.
+      operands.removeIf(x -> x.equals(FALSE));
+      if (operands.size() == 0) {
+        return setExpressionToBoolean(expression, false);
+      }
+    }
+
+    return expression;
+  }
+
+  private boolean isNumericColumn(Expression expression, Schema schema) {
+    if (expression.getType() != ExpressionType.IDENTIFIER) {
+      // Expression can not be a column.
+      return false;
+    }
+
+    String column = expression.getIdentifier().getName();
+    FieldSpec fieldSpec = schema.getFieldSpecFor(column);
+    if (fieldSpec == null || !fieldSpec.isSingleValueField()) {
+      // Expression can not be a column name.
+      return false;
+    }
+
+    return schema.getFieldSpecFor(column).getDataType().isNumeric();
+  }
+
+  private boolean isNumericLiteral(Expression expression) {
+    if (expression.getType() == ExpressionType.LITERAL) {
+      Literal._Fields type = expression.getLiteral().getSetField();
+      switch (type) {
+        case SHORT_VALUE:
+        case INT_VALUE:
+        case LONG_VALUE:
+        case DOUBLE_VALUE:
+          return true;
+      }
+    }
+    return false;
+  }
+
+  /** Change the expression value to boolean literal with given value. */
+  private static Expression setExpressionToBoolean(Expression expression, boolean value) {
+    expression.unsetFunctionCall();
```

Comment:
This seems hacky. We should simply create a new one instead of reusing the existing one

Reply:
Same here, all that this function is doing is modifying the reference that it is getting which is again a very standard thing to do.

Reply:
I understand the intention of this function. What I'm saying is turning a function expression into a literal expression by unsetting the function and then setting a literal is not necessary (it won't be correct if expression is an identifier), instead we can simply return a constant boolean literal here which is more readable.